Fixes #804 Json parsing errors mention JSON.parse in errors

nojvek opened the pull request:
I added JSON.parse specific errors. It still doesn't feel very clean as there is no column info. Would this be relatively easy to add?

msftclas commented:
Hi __@nojvek__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Manoj Patel). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

ianwjhalliday commented:
Thanks @nojvek, these new error messages are definitely more helpful.

The error stack will have the line column information for the source of the throw, which will be where the JSON.parse call occurred, so we do have that already.

Could also be useful to have character offset (or line/column?) information for the source of the error in the JSON string passed to parse.  I don't know if that would require much additional work on in the JSON parser or not.  Certainly this can be raised as a new separate `Suggestion` issue if you think it is a good idea.

@abchatra any concerns over adding new error messages?

nojvek commented:
Right now there seems to be a LSC_ERROR_MSG(..) macro definition of a
syntax error.

How do I throw a syntax error with a dynamically created string ?



On Wed, May 18, 2016 at 2:03 PM, Ian Halliday <notifications@github.com>
wrote:

> Thanks @nojvek <https://github.com/nojvek>, these new error messages are
> definitely more helpful.
>
> The error stack will have the line column information for the source of
> the throw, which will be where the JSON.parse call occurred, so we do have
> that already.
>
> Could also be useful to have character offset (or line/column?)
> information for the source of the error in the JSON string passed to parse.
> I don't know if that would require much additional work on in the JSON
> parser or not. Certainly this can be raised as a new separate Suggestion
> issue if you think it is a good idea.
>
> @abchatra <https://github.com/abchatra> any concerns over adding new
> error messages?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/Microsoft/ChakraCore/pull/1001#issuecomment-220157158>
>

ianwjhalliday commented:
Look at `rterrors.h` where we have errors with two forms of their message string.  The first form has a %s parameter that will be replaced by the given string to the `JavascriptError::Throw*` methods.

I believe you should be able to move the errors to `rterrors.h` and still use them for `ThrowSyntaxError` without any issues.

abchatra commented:
> @abchatra any concerns over adding new error messages?

I am happy to see improvement in the error messages.

nojvek commented:
Thanks abchatra for the help today. I have made the fixes to show the position just like v8 does for JSON.parse. wdyt?

ianwjhalliday commented:
Looks great!  Thanks @nojvek, I'll get this merged.

ianwjhalliday commented:
@nojvek turns out we have some JSON tests that do not run in jenkins currently because they depend on timezone.

You will need to manually run these two tests, `test/JSON/jx2.js` and `test/es5/jx2.js`, and update their baselines to match your change before we can merge this in.  These two tests happen to run with no ch.exe command line switches, but in general when running tests manually you'd want to confirm you are using the right switches by examining the `test/.../rlexe.xml` file to see if it specifies switches and if so what they are.  Also this will show you what the baseline file is named, sometimes it doesn't match the source file.

I also just noticed that your new test `syntaxerror.js` is not added to `test/JSON/rlexe.xml`.  Please add it there so that it runs in the test runner.

ianwjhalliday commented:
Note the json tests that don't run in jenkins because of their timezone dependency is an understood issue that we have not fixed yet.  It depends on PST timezone.

abchatra commented on lib/Runtime/Library/JSONParser.cpp:
```diff
@@ -404,12 +404,13 @@ namespace JSON
         default:
-            Js::JavascriptError::ThrowSyntaxError(scriptContext, ERRsyntax);
+            m_scanner.ThrowSyntaxError(JSERR_JsonSyntax);
+            return NULL;
```
Not required. This statement won't be reached.

nojvek replied:
Without this C++ won't compile.

>e:\personal\chakracore\lib\runtime\library\jsonparser.cpp(415): warning C4715: 'JSON::JSONParser::ParseObject': not all control paths return a value

ianwjhalliday replied:
That is because `JSONScanner::ThrowSyntaxError` is not marked with `__declspec(noreturn)`.

Probably reasonable to change the definition of `JSONScanner::ThrowSyntaxError` to:
```c++
void __declspec(noreturn) ThrowSyntaxError(int wErr)
```

nojvek replied:
done.

abchatra commented on lib/Runtime/Library/JSONScanner.h:
```diff
@@ -20,9 +20,16 @@ namespace JSON
+        void ThrowSyntaxError(int wErr)
+        {
+            char16 scanPos[16];
+            ::_itow_s(GetScanPos(), scanPos, sizeof(scanPos) / sizeof(char16), 10);
```
Use _countof(scanPos)

nojvek replied:
done.

abchatra commented on lib/Runtime/Library/JSONScanner.h:
```diff
@@ -20,9 +20,16 @@ namespace JSON
-        uint GetCurrentStringLen(){return currentIndex;}
+        char16* GetCurrentString() { return currentString; } 
+        uint GetCurrentStringLen() { return currentIndex; }
+        uint GetScanPos() { return uint(currentChar - inputText); }
```
Nit: Rename to GetScanPosition?

nojvek replied:
done.

abchatra commented on lib/Runtime/Library/JSONScanner.h:
```diff
@@ -20,9 +20,16 @@ namespace JSON
+        uint GetCurrentStringLen() { return currentIndex; }
+        uint GetScanPos() { return uint(currentChar - inputText); }
 
+        void ThrowSyntaxError(int wErr)
```
This should be long right?

nojvek replied:
the function takes a int32 hcode

err_method(ScriptContext* scriptContext, int32 hCode, PCWSTR varName = nullptr)

nojvek commented:
Ensured tests pass in  runtests.cmd -x64 -debug and added syntaxError.js to rl.xml

ianwjhalliday commented:
@nojvek you forgot to update the baseline for `test\es5\jx2.js` as well.

nojvek commented:
Oh forgot that. I updated Json/jx2.js

On Wednesday, June 1, 2016, Ian Halliday <notifications@github.com> wrote:

> @nojvek <https://github.com/nojvek> you forgot to update the baseline for
> test\es5\jx2.js as well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/ChakraCore/pull/1001#issuecomment-223145333>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA-JVFQNPXjf5SZyJzGzCqVO7-G9Su0vks5qHgkFgaJpZM4Ig4Jc>
> .
>

ianwjhalliday commented:
@nojvek when you get a chance please update `test\es5\jx3.baseline`.  I've updated internal test baselines so once you've done that I believe we'll be all ready to merge your change.

nojvek commented:
Done. Ran all 1655 tests and confirmed they passed.

Not sure why baseline for jx2.js was called jx3.js. I've renamed to
jx2.baseline.

Soz for delay

On Thursday, June 2, 2016, Ian Halliday <notifications@github.com> wrote:

> @nojvek <https://github.com/nojvek> when you get a chance please update
> test\es5\jx3.baseline. I've updated internal test baselines so once
> you've done that I believe we'll be all ready to merge your change.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/ChakraCore/pull/1001#issuecomment-223445011>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA-JVKI0UcnShH3KDFscE4dJsXbvpJkKks5qH10xgaJpZM4Ig4Jc>
> .
>

ianwjhalliday commented:
Awesome, thanks for the rename.  I don't know why it was different either.  Probably some historical reason that didn't get cleaned up nicely.

nojvek commented:
Would be great to see this go in today. Let me know if you need anything else from my side. Thanks

ianwjhalliday commented:
@nojvek we're all good, merged!  Thanks for improving these error messages

nojvek commented:
That's awesome. Thanks Ian.

On Mon, Jun 6, 2016 at 2:18 PM, Ian Halliday <notifications@github.com>
wrote:

> @nojvek <https://github.com/nojvek> we're all good, merged! Thanks for
> improving these error messages
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/ChakraCore/pull/1001#issuecomment-224091524>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA-JVDjgc-Yok9s9XJygkux6ofi-Wcupks5qJI60gaJpZM4Ig4Jc>
> .
>

abchatra commented:
Thank you @nojvek

dilijev commented:
Note that we're tracking the timezone-sensitive tests issue in #319
